### PR TITLE
Adding Xi-compatible tracks to lostTracks collection

### PR DESCRIPTION
**Pull request by AdrianoDee**
#### PR description:

This PR adds to lostTracks collection all those tracks (pions) that can potentially form Λπ candidate checking the invariant mass M( Λπ)< X GeV. This PR adds to the lostTracks modules a variable `xiSelection` to turn on this selection and a tunable mass cut `(xiMassCut `).

#### PR validation:

All the tests run. 

The size increase is <~ 1% evaluated on JetHT Run2018D data with <PU>=37

See slide 10 [here](https://indico.cern.ch/event/1042169/contributions/4383052/attachments/2254247/3824630/trk4bph-RECO.pdf)

### CODE REVIEW

File: PhysicsTools/PatAlgos/plugins/PATLostTracks.cc [hunk 1]
```diff
@@ -190,10 +196,28 @@ void pat::PATLostTracks::produce(edm::StreamID, edm::Event& iEvent, const edm::E
...
+    double protonCharge = 0;
     for (size_t dIdx = 0; dIdx < v0.numberOfDaughters(); dIdx++) {
       size_t key = (dynamic_cast<const reco::RecoChargedCandidate*>(v0.daughter(dIdx)))->track().key();
       if (trkStatus[key] == TrkStatus::NOTUSED)
         trkStatus[key] = TrkStatus::VTX;
+      protonCharge += v0.daughter(dIdx)->charge() * v0.daughter(dIdx)->momentum().mag2();
+    }
+    if (xiSelection_) {
+      // selecting potential Xi- -> Lambda pi candidates
+      TLorentzVector p4Lambda;
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATLostTracks.cc:208]: please switch to a `math::` Lorentz vector. Given the cost of computing eta and phi for tracks, I think that `math::XYZTLorentzVector` is more appropriate.

**Review by slava77 — COMMENTED** [review]
please provide a timing measurement, preferably using igprof. The nested loop over tracks may be quite costly and then will need to be rewritten to minimize recomputations.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by AdrianoDee** [comment]
> please provide a timing measurement, preferably using igprof. The nested loop over tracks may be quite costly and then will need to be rewritten to minimize recomputations.

Sure, about this it's worth to move the

`if (trkStatus[trkIndx] == TrkStatus::NOTUSED)`

above before the p4 calculations. With this implemented the igprof reports:

- [original]( https://adiflori.web.cern.ch/cgi-bin/igprof-navigator/old_perf/1996)
- [new](https://adiflori.web.cern.ch/cgi-bin/igprof-navigator/new_perf/1534)

The timing for the module goes from ~150 ms per 100 events to 220 ms per 100 events.

**Comment by slava77** [comment]
> > please provide a timing measurement, preferably using igprof. The nested loop over tracks may be quite costly and then will need to be rewritten to minimize recomputations.
> 
> Sure, about this it's worth to move the
> 
> `if (trkStatus[trkIndx] == TrkStatus::NOTUSED)`
> 
> above before the p4 calculations. With this implemented the igprof reports:
> 
>     * [original](https://adiflori.web.cern.ch/cgi-bin/igprof-navigator/old_perf/1996)
> 
>     * [new](https://adiflori.web.cern.ch/cgi-bin/igprof-navigator/new_perf/1534)
> 
> 
> The timing for the module goes from ~150 ms per 100 events to 220 ms per 100 events.

the igprof shows that the cost of the loop in question is reasonably small and further optimization is not required.
 IIUC, this still requires a change in the location of `if (trkStatus[trkIndx] == TrkStatus::NOTUSED)` compared to the current state of this PR

**Comment by AdrianoDee** [comment]
> IIUC, this still requires a change in the location of `if (trkStatus[trkIndx] == TrkStatus::NOTUSED)` compared to the current state of this PR

Indeed, updated it with the new setup

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
urgent
in the context of getting this to 106X for bParking mini

**Comment by mtosi** [comment]
thanks a lot for this very fast review (really appreciated !)
is everything fine, or are there still open items ?

thanks, and have a nice day (and Festa della Repubblica for the Italians ;) )

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33891	807d8e0
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show differences in quantities related to the lostTracks
    - DQM plots show more entries in /lostTracks histograms, as expected
    - CPU cost seems reasonable based on igprofs provided earlier (the extra cost does not appear significant in the selection loop, while most of the added cost is in creation of the lost track packed candidates)
        * [original](https://adiflori.web.cern.ch/cgi-bin/igprof-navigator/old_perf/1996)
        * [new](https://adiflori.web.cern.ch/cgi-bin/igprof-navigator/new_perf/1534)
    - storage cost is in agreement with the PR description: in wf 136.88811 the output file is up by 0.26%

the wf 136.88811 shows about 26% increase in the number of lost tracks, mostly at low pt
all_OldVSNew_RunJetHT2018DreMINIAODULwf136p88811c_log10patPackedCandidates_lostTracks__PAT_obj_pt

**Comment by qliphy** [comment]
+1

**Comment by AdrianoDee** [comment]
Thanks!

**Pull request merged**